ui/client/reducers/toolbars.ts

```diff
@@ -0,0 +1,117 @@
+import {Reducer} from "../actions/reduxTypes"
+import {combineReducers} from "redux"
+
+export enum ToolbarsSide {
+  TopRight = "TOP-RIGHT",
+  BottomRight = "BOTTOM-RIGHT",
+  TopLeft = "TOP-LEFT",
+  BottomLeft = "BOTTOM-LEFT",
+  Hidden = "HIDDEN",
+}
+
+type NodeToolbox = {
+  opened: Record<string, boolean>,
+}
+
+type InitData = Array<[string, ToolbarsSide]>
+
+type Positions = {
+  [side in ToolbarsSide]?: string[]
+}
+
+type Collapsed = Record<string, boolean>
+
+export type ToolbarsState = {
+  positions: Positions,
+  initData: InitData,
+  collapsed: Collapsed,
+  nodeToolbox: NodeToolbox,
+}
+
+const nodeToolbox: Reducer<NodeToolbox> = (state = {opened: {}}, action) => {
+  switch (action.type) {
+    case "TOGGLE_NODE_TOOLBOX_GROUP":
+      return {
+        opened: {
+          ...state.opened,
+          [action.nodeGroup]: !state.opened[action.nodeGroup],
+        },
+      }
+
+    case "RESET_TOOLBARS":
+      return {
+        opened: {},
+      }
+
+    default:
+      return state
+  }
+}
+
+function setupPositions(positions: Positions, toolbars: Array<[string, ToolbarsSide]>): Positions {
+  const groups = Object.values(positions)
+  const newToolbars = toolbars.filter(([id]) => !groups.some(g => g.includes(id)))
+  return newToolbars.reduce((nextState, [id, side]) => {
+    const currentValues = nextState[side || ToolbarsSide.TopRight] || []
```

Comment:
What happen here? Why `TopRight` is special? Is it some kind of logic iterating clockwise and TopRight is a "0 hour"?

Reply:
ok, I know now that it is only change for right panel and it is default value